Core/include/Acts/TrackFinder/CombinatorialKalmanFilter.hpp

```diff
@@ -0,0 +1,1255 @@
+                  << filteredParams.parameters().transpose()
+                  << " of track state with tip = " << result.currentTip);
+            }
+          }
+        }
+        // Update state and stepper with post material effects
+        materialInteractor(surface, state, stepper, postUpdate);
+      } else if (surface->surfaceMaterial() != nullptr) {
+        // Retrieve the tip state and remove the current tip from active tips
+        TipState preTipState;
+        auto tip_it = result.activeTips.find(result.currentTip);
+        if (tip_it != result.activeTips.end()) {
+          preTipState = tip_it->second;
+          result.activeTips.erase(tip_it);
+        }
+
+        // Create state if there is already measurement detected on this branch
+        if (preTipState.nMeasurements > 0) {
+          // No source links on surface, add either hole or passive material
+          // TrackState. No storage allocation for
+          // uncalibrated/calibrated measurement and filtered parameter
+          result.currentTip = result.fittedStates.addTrackState(
+              ~(TrackStatePropMask::Uncalibrated |
+                TrackStatePropMask::Calibrated | TrackStatePropMask::Filtered),
+              result.currentTip);
+
+          ACTS_VERBOSE("Creating non-sourcelink track state with tip = "
+                       << result.currentTip);
+
+          // Count the number of processedStates, measurements, outliers and
+          // holes
+          auto tipState =
+              TipState{preTipState.nStates + 1, preTipState.nMeasurements,
+                       preTipState.nOutliers, preTipState.nHoles};
+
+          // now get track state proxy back
+          auto trackStateProxy =
+              result.fittedStates.getTrackState(result.currentTip);
+
+          // Set the track state flags
+          auto& typeFlags = trackStateProxy.typeFlags();
+          typeFlags.set(TrackStateFlag::MaterialFlag);
+          typeFlags.set(TrackStateFlag::ParameterFlag);
+
+          if (surface->associatedDetectorElement() != nullptr) {
+            ACTS_VERBOSE("Detected hole on " << surface->geoID());
+            // If the surface is sensitive, set the hole type flag
+            typeFlags.set(TrackStateFlag::HoleFlag);
+
+            // Increment of number of holes
+            tipState.nHoles++;
+
+            // Transport & bind the state to the current surface
+            auto [boundParams, jacobian, pathLength] =
+                stepper.boundState(state.stepping, *surface, true);
+
+            // Set the surface
+            trackStateProxy.setReferenceSurface(surface->getSharedPtr());
+
+            // Fill the track state
+            trackStateProxy.predicted() = boundParams.parameters();
+            trackStateProxy.predictedCovariance() = *boundParams.covariance();
+            trackStateProxy.jacobian() = jacobian;
+            trackStateProxy.pathLength() = pathLength;
+          } else {
```

Comment:
So this would be a pure *material surface*, right?

Reply:
Exactly.